### PR TITLE
Add documented way to pass data to custom steps

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

File: src/linker/Linker.Steps/ClearInitLocals.cs [hunk 1]
```diff
@@ -12,8 +12,7 @@ protected override void Process ()
 		{
 			string parameterName = "ClearInitLocalsAssemblies";
```
Comment (vitek-karas) [inline src/linker/Linker.Steps/ClearInitLocals.cs:13]:
```suggestion
```

File: src/linker/Linker.Steps/ClearInitLocals.cs [hunk 2]
```diff
@@ -12,8 +12,7 @@ protected override void Process ()
...
 
-			if (Context.HasParameter (parameterName)) {
-				string parameter = Context.GetParameter (parameterName);
+			if (Context.TryGetCustomData (parameterName, out string parameter)) {
```
Comment (vitek-karas) [inline src/linker/Linker.Steps/ClearInitLocals.cs:15]:
```suggestion
			if (Context.TryGetCustomData ("ClearInitLocalsAssemblies", out string parameter)) {
```

File: src/linker/README.md [hunk 3]
```diff
@@ -175,6 +175,11 @@ Or before a specific step:
 
 `illink --custom-step -MarkStep:Foo.FooStep,D:\Bar\Foo.dll -a program.exe`
 
+### Passing data to custom steps
+
+For advanced custom steps which needs interaction with external values (for example for the custom step configuration), there is `--custom-data KEY=VALUE` option. The key
+data are stored inside a linker context and can be obtained in the custom stop using `context.TryGetCustomData` method. Each key can have a simple value assigned which means
```
Comment (vitek-karas) [inline src/linker/README.md:181]:
```suggestion
data are stored inside a linker context and can be obtained in the custom step using `context.TryGetCustomData` method. Each key can have a simple value assigned which means
```

File: test/Mono.Linker.Tests.Cases/CommandLine/Dependencies/CustomStepUser.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+using Mono.Linker.Steps;
+
+namespace CustomStep
+{
+	public class CustomStepUser : IStep
+	{
+		public void Process(LinkContext context)
+		{
+			if (context.TryGetCustomData ("NewKey", out var value))
+				context.LogMessage ("Custom step added with custom data of " + value);
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/CommandLine/Dependencies/CustomStepUser.cs:12]: This should throw if the value is not the expected `UserValue` - printing it out doesn't actually verify the value in any way.

Comment (marek-safar) [reply test/Mono.Linker.Tests.Cases/CommandLine/Dependencies/CustomStepUser.cs:12]: The value is verified at https://github.com/mono/linker/pull/1128/files#diff-26f773e2a62fedeb70de25da1d59505cR15. Do you think that's not enough?

Comment (vitek-karas) [reply test/Mono.Linker.Tests.Cases/CommandLine/Dependencies/CustomStepUser.cs:12]: Sorry - I missed that part - looks good.

**Pull request merged**